providers/okta/okta_auth.go

```diff
@@ -0,0 +1,205 @@
+	return factors
+}
+
+func (p AuthResponse) UserID() string {
+	var user struct {
+		ID string `json:"id"`
+	}
+
+	// Error intentionally ignored.
+	p.Embedded.Get("user", &user)
+	return user.ID
+}
+
+func NewAuthClient(hostname string, opts ...func(*AuthClient)) AuthClient {
+	client := AuthClient{
+		url: url.URL{
+			Scheme: "https",
+			Host:   hostname,
+		},
+		rt: http.DefaultClient,
+	}
+
+	for _, opt := range opts {
+		opt(&client)
+	}
+
+	return client
+}
+
+func WithHTTPClient(client *http.Client) func(*AuthClient) {
+	return func(a *AuthClient) {
+		a.rt = client
+	}
+}
+
+func (o AuthClient) do(ctx context.Context, method, path string, data, result interface{}) (response *http.Response, err error) {
+	b, err := json.Marshal(data)
+	if err != nil {
+		err = wrapError(err, ErrOktaBadRequest)
+		return
+	}
+
+	uri := o.url
+	uri.Path = path
+	req, err := http.NewRequestWithContext(ctx, method, uri.String(), bytes.NewReader(b))
+	if err != nil {
+		err = wrapError(err, ErrOktaBadRequest)
+		return
+	}
+
+	req.Header.Set("content-type", "application/json")
+	req.Header.Set("accept", "application/json")
+
+	response, err = o.rt.Do(req)
+	if err != nil {
+		return
+	}
+
+	dec := json.NewDecoder(response.Body)
+	defer response.Body.Close()
+
+	switch response.StatusCode {
+	case http.StatusOK:
+		err = dec.Decode(result)
+		return
+	default:
+		var inner okta.Error
+		err = dec.Decode(&inner)
+		if err != nil {
+			return
+		}
+
+		err = &inner
+		return
+	}
+}
+
+// Authn posts a request to the /authn endpoint.
+func (o AuthClient) VerifyCredentials(ctx context.Context, payload AuthRequest) (AuthResponse, error) {
+	var res AuthResponse
+	httpResponse, err := o.do(ctx, "POST", "/api/v1/authn", payload, &res)
+	return res, translateError(httpResponse, err)
+}
+
+func getHrefLink(app okta.Application) (string, bool) {
+	links := app.Links.(map[string]interface{})
+	appLinks := links["appLinks"].([]interface{})
+	for _, interf := range appLinks {
+		entry := interf.(map[string]interface{})
+		if entry["type"] == "text/html" {
+			return entry["href"].(string), true
+		}
+	}
+
+	return "", false
+}
+
+// OktaError is an error from Okta.
+type OktaError error
+
+// A list of standard errors that can be returned by by Okta client.
+var (
+	ErrOktaBadRequest                   OktaError = errors.New("bad request")
+	ErrOktaUnauthorized                 OktaError = errors.New("unauthorized")
+	ErrOktaForbidden                    OktaError = errors.New("forbidden")
+	ErrOktaCouldNotStepUpAuthentication OktaError = errors.New("could not create a session")
+	ErrOktaSAMLError                    OktaError = errors.New("could not get a SAML response")
+	ErrOktaInternalServerError          OktaError = errors.New("internal server error")
+	ErrOktaUnspecified                  OktaError = errors.New("unspecified")
+)
+
+// wrapError wraps an error into a standard Okta client error.
+func wrapError(err error, oktaErr OktaError) error {
```

Comment:
I'm still not sure of the utility of this.

If we really want our own error type, it should probably look something like this:

```go
type OktaError struct {
    Inner error
    Code string
}

func (o OktaError) Unwrap() error {
    return o.Inner
}

func (o OktaError) Error() string {
    return fmt.Sprintf("%s: %s", o.Inner.Error(), o.Code)
}
```

Reply:
yeah agreed yours is definitely more go friendly pattern